aidb/vector_database/marqo_vector_database.py

```diff
@@ -0,0 +1,223 @@
+import uuid
+from dataclasses import dataclass, field
+import numpy as np
+import pandas as pd
+import marqo
+from typing import Dict,  Optional
+
+from aidb.utils.logger import logger
+from aidb.vector_database.vector_database import VectorDatabase
+
+@dataclass
+class MarqoAuth:
+  """
+  :param url: marqo url
+  :param username: marqo username
+  :param pwd: marqo password
+  :param api_key: marqo api key, user should choose input either username/pwd or api_key
+  """
+  url: Optional[str] = field(default=None)
+  username: Optional[str] = field(default=None)
+  pwd: Optional[str] = field(default=None)
+  api_key: Optional[str] = field(default=None)
+
+class MarqoVectorDatabase(VectorDatabase):
+  def __init__(self, marqo_auth: MarqoAuth, embedding_dim: int) -> None:
+    '''
+    Authentication
+    '''
+    if marqo_auth.url is None:
+      raise ValueError('Marqo requires URL to connect')
+    
+    self.marqo_client = marqo.Client(
+      url=marqo_auth.url, main_user=marqo_auth.username, main_password=marqo_auth.pwd)
+    status = self.marqo_client.get_marqo()
+    if not status:
+      raise Exception('Initial connection to Marqo failed')
+    
+    self.index_list = [i.index_name for i in self.marqo_client.get_indexes()['results']]
+    self.embedding_dim = embedding_dim
```

Comment:
better to put this in 'create_index'

Reply:
Okay I would move it, actually I saw in weaviate that it was in `__init__` and so did it.